### PR TITLE
Support for RSK

### PR DESCRIPTION
**Pull request by philipglazman**
Preliminary PR for Rootstock support in SDK.

Added support for Rootstock wallet creation, addresses, and transaction building/signing.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -4,3 +4,6 @@
 [submodule "vendor/asmcrypto.js"]
 	path = vendor/asmcrypto.js
 	url = https://github.com/vibornoff/asmcrypto.js
+[submodule "vendor/ethereumjs-wallet"]
```
Comment (afk11) [inline .gitmodules]: why submodule this in?

File: lib/api_client.js [hunk 2]
```diff
@@ -1277,34 +1284,64 @@ APIClient.prototype.initWallet = function(options, cb) {
                 throw new Error("Backup key returned from server didn't match our own copy");
             }
         }
-        var backupPublicKey = bitcoin.HDNode.fromBase58(result.backup_public_key[0], self.network);
-        var blocktrailPublicKeys = _.mapValues(result.blocktrail_public_keys, function(blocktrailPublicKey) {
-            return bitcoin.HDNode.fromBase58(blocktrailPublicKey[0], self.network);
-        });
-        var primaryPublicKeys = _.mapValues(result.primary_public_keys, function(primaryPublicKey) {
+
+        var backupPublicKey,blocktrailPublicKeys,primaryPublicKeys, wallet;
```
Comment (afk11) [inline lib/api_client.js]: spacing nit

File: lib/wallet.js [hunk 3]
```diff
@@ -695,6 +709,80 @@ Wallet.prototype.upgradeKeyIndex = function(keyIndex, cb) {
...
+            // was a variable but not integer
+            deferred.reject(new Error("Invalid chain index"));
+            return deferred.promise;
+        }
+    }
+
+    // Get Rootstock address from server.
+    deferred.resolve(self.sdk.getNewDerivation(self.identifier, "M/" + self.keyIndex + "'/" + chainIdx)
+        .then(function(newDerivation) {
+            var path = newDerivation.path.replace("\'","").replace('9999',"9999'");
```
Comment (afk11) [inline lib/wallet.js]: not sure this behaves as intended - it seems to remove all `'`s, placing in 9999' wherever 9999 is encountered, which can also be found in the level we use for address derivations

what's it supposed to do? is it necessary?

File: lib/api_client.js [hunk 4]
```diff
@@ -2129,6 +2215,40 @@ APIClient.prototype.sendTransaction = function(identifier, txHex, paths, checkFe
...
+        prioboost = false;
+    } else if (typeof options === "function") {
+        cb = options;
+        options = {};
+    }
+
+    var data = {
+        two_factor_token: twoFactorToken,
+        raw_transaction: txHex,
+        transaction_id: txid
```
Comment (afk11) [inline lib/api_client.js]: naming nit re transaction_id. was this required for the btc style txs? might be a legacy thing (reviewing nodejs sdk first though)

File: lib/wallet.js [hunk 5]
```diff
@@ -1886,4 +1974,220 @@ Wallet.deriveByPath = function(hdKey, path, keyPath) {
...
+                        cb();
+                    },
+                    /**
+                     * add transaction params
+                     *
+                     * @param cb
+                     */
+                    function(cb) {
+                        
+                        //@TODO - implement safety check for nonce
```
Comment (afk11) [inline lib/wallet.js]: todo

File: lib/wallet.js [hunk 6]
```diff
@@ -1886,4 +1974,220 @@ Wallet.deriveByPath = function(hdKey, path, keyPath) {
...
+            function(r) { deferred.resolve(r); },
+            function(e) { deferred.reject(e); }
+        )
+        ;
+    });
+
+    return deferred.promise;
+};
+
+Wallet.prototype.buildRskTransaction = function(pay,feeStrategy, options, cb) {
```
Comment (afk11) [inline lib/wallet.js]: missing documentation for pay object structure. consider breaking into arguments if it's really small?

File: package.json [hunk 7]
```diff
@@ -88,7 +94,7 @@
     "browserify": "*",
     "browserify-versionify": "^1.0.6",
     "coveralls": "^2.13.1",
-    "grunt": "~0.4.2",
+    "grunt": "^0.4.2",
```
Comment (afk11) [inline package.json]: guessing it rolled up to 0.7 or something crazy. this is definitely the conservative option..

**Comment by afk11** [comment]
we normally avoid committing changes to the build directory unless it's build time - since it's between tags there's no real need to update it

**Review by afk11 — COMMENTED** [review]
looking good! noticed there isn't a createNewWallet update for this, does it just work out?

File: .gitmodules (hunk 1 above)
Comment (MangoSalad) [reply .gitmodules]: I remember at the time that I could not find an equivalent npm package. Removed it and will use npm to include ethereumjs-wallet.

File: lib/api_client.js (hunk 2 above)
Comment (MangoSalad) [reply lib/api_client.js]: thanks - this is helpful.

File: lib/wallet.js (hunk 3 above)
Comment (MangoSalad) [reply lib/wallet.js]: Right, removing this as the path is parsed later on in Wallet.deriveByPath. This is a relic from testing.

File: lib/api_client.js (hunk 4 above)
Comment (MangoSalad) [reply lib/api_client.js]: It's not required for broadcasting. But backend does return txid in a view.

**Comment by MangoSalad** [comment]
No need for a createNewWallet since it builds upon an existing wallet and uses separate code paths if the initialised wallet is using the rootstock network.

**Comment by afk11** [comment]
sorry! https://github.com/blocktrail/blocktrail-sdk-nodejs/blob/2404e55d8b4b1d62ac2ea5c452f47d006f39c6ea/lib/wallet.js#L889 maybe just revert to the array shizzle, seems we keep to that elsewhere

File: lib/api_client.js [hunk 8]
```diff
@@ -184,7 +198,7 @@ APIClient.normalizeNetworkFromOptions = function(options) {
 
     apiNetwork = (prefix || "") + network;
 
-    return [network, testnet, regtest, apiNetwork];
+    return [network, testnet, regtest, apiNetwork, rskTestnet];
```
Comment (afk11) [inline lib/api_client.js]: hmm, any reason we can't use testnet without adding rskTestnet?

**Pull request closed without merging**